walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,88 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors; 
+import java.util.stream.Stream; 
 
 public final class FindMeetingQuery {
+  /**
+   * Method for retrieving all possible times for the request based on all attendee's
+   * events and availabilty.
+   * @return Collection of TimeRanges that show when the requested meeting can occur.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    int dayStart = TimeRange.START_OF_DAY;
+    int dayEnd = TimeRange.END_OF_DAY;
+    int startEvent = dayStart;
+    int endEvent;
+    
+    TimeRange previousEventTimeRange = TimeRange.fromStartDuration(0,0);  
+    List<TimeRange> sortedEventTimeRanges = getAllAttendeeRanges(events, request);
+    ArrayList<TimeRange> availableTimeRangesForRequest = new ArrayList<>();
+
+    //Returns an empty List should the requested duration be longer than a total day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) return availableTimeRangesForRequest;
+
+    //For-each that iterates through sortedTimeRanges to determine available times for request's duration.
+    for (TimeRange timeRange : sortedEventTimeRanges) {
+        if (timeRange.start() == dayStart) startEvent = timeRange.end();
+        //If the gap between the current start and the next event start is greater than request's duration
+        //create a TimeRange for that period.
+        if (timeRange.start() - startEvent >= request.getDuration()) {
+            availableTimeRangesForRequest.add(TimeRange.fromStartEnd(startEvent, timeRange.start(), false));
+        }
+        startEvent = timeRange.end();
+        endEvent = timeRange.end();
+      
+        //Checks to see if the TimeRanges are nested.
+        if (previousEventTimeRange.contains(timeRange)) {
+            endEvent = previousEventTimeRange.end();
+        }
+      
+        //If the current TimeRange is the last in the list then determine if there is space for the request after.
+        if (sortedEventTimeRanges.get(sortedEventTimeRanges.size() - 1) == timeRange) {
```

Comment:
this is a very usage of == comparing object reference.
I guess since we need to know if we reaches end of the list, we can use the iterator. https://stackoverflow.com/questions/41591107/detect-last-foreach-loop-iteration